File: frontend/src/api/schedules.ts
import { apiFetch } from './client';
import type { Page, ScanDetail, ScanParams } from './scans';

export interface BaselineRule {
  mode: 'same_season_previous_year' | 'years_back' | 'fixed';
  window_days?: number | null;
  years?: number | null;
  baseline_start?: string | null;
  baseline_end?: string | null;
}

export interface Schedule {
  id: string;
  name: string;
  is_active: boolean;
  cron: string;
  current_window_days: number;
  baseline_rule: BaselineRule;
  params: ScanParams;
  last_run_at: string | null;
  next_run_at: string | null;
  created_by: string | null;
  created_at: string;
  parcels: { id: string; name: string; category: string }[];
  last_scan_id: string | null;
  last_scan_status: string | null;
}

export interface ScheduleCreate {
  name: string;
  parcel_ids: string[];
  cron: string;
  current_window_days: number;
  baseline_rule: BaselineRule;
  params?: Partial<ScanParams>;
}

export type SchedulePatch = Partial<ScheduleCreate> & { is_active?: boolean };

export const listSchedules = () => apiFetch<Page<Schedule>>('schedules?page_size=100');
export const createSchedule = (body: ScheduleCreate) =>
  apiFetch<Schedule>('schedules', { method: 'POST', body: JSON.stringify(body) });
export const patchSchedule = (id: string, body: SchedulePatch) =>
  apiFetch<Schedule>(`schedules/${id}`, { method: 'PATCH', body: JSON.stringify(body) });
export const runSchedule = (id: string) =>
  apiFetch<ScanDetail>(`schedules/${id}/run`, { method: 'POST' });
export const deleteSchedule = (id: string) =>
  apiFetch<void>(`schedules/${id}`, { method: 'DELETE' });

File: backend/app/schemas/schedules.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.scans import ScanParamsIn, ScanParcelOut


class BaselineRule(BaseModel):
    mode: Literal["same_season_previous_year", "years_back", "fixed"] = "same_season_previous_year"
    window_days: int | None = Field(default=None, ge=5, le=120)
    years: int | None = Field(default=None, ge=1, le=8)
    baseline_start: str | None = None  # ISO date, mode=fixed
    baseline_end: str | None = None


class ScheduleCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    parcel_ids: list[uuid.UUID] = Field(min_length=1, max_length=500)
    cron: str = Field(min_length=1, max_length=100, description="'weekly' | 'monthly' | crontab")
    current_window_days: int = Field(default=30, ge=5, le=120)
    baseline_rule: BaselineRule = Field(default_factory=BaselineRule)
    params: ScanParamsIn = Field(default_factory=ScanParamsIn)


class SchedulePatch(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    parcel_ids: list[uuid.UUID] | None = Field(default=None, min_length=1, max_length=500)
    cron: str | None = Field(default=None, min_length=1, max_length=100)
    current_window_days: int | None = Field(default=None, ge=5, le=120)
    baseline_rule: BaselineRule | None = None
    params: ScanParamsIn | None = None
    is_active: bool | None = None


class ScheduleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    is_active: bool
    cron: str
    current_window_days: int
    baseline_rule: dict[str, Any]
    params: dict[str, Any]
    last_run_at: datetime | None
    next_run_at: datetime | None
    created_by: uuid.UUID | None
    created_at: datetime
    parcels: list[ScanParcelOut] = []
    last_scan_id: uuid.UUID | None = None
    last_scan_status: str | None = None
